### PR TITLE
Stridearrays 0.4 updates

### PR DESCRIPTION
**Pull request by chriselrod**
This adds further updates for the breaking changes of StrideArraysCore 0.4.

A great deal of the breakages were also that things were outright broken in StrideArraysCore; the author of that library ought to rethink some of their decision making to consider if such changes were really worthwhile.

Regardless, I've hunted down a lot of bugs there, as well as subtleties within this library.

It may still be worth rethinking at some point how to simplify things, as there is probably too much to consider that is undocumented, particularly when coming back after not looking at the code for a few months.

Or at least, not enough checking and not enough tooling.

The change to StrideArraysCore was to avoid storing unnecessary stride information.
Earlier versions held a pointer, a size tuple, and a stride tuple.
That stride tuple is now "sparse", with `nothing` entries whenever that stride equals the product of the previous stride and size, where previous means "next lowest rank", i.e. it has to support arbitrary permutations of axis, in which case it tracks the relative order.

That was a mostly internal change, and almost all fixes were within StrideArraysCore itself, dealing with bugs there.
In particular, with `BitPtrArray`, i.e. arrays of packed bits. SimpleChains uses these, makes views, etc, and this wasn't well tested.

The diff here in SimpleChains should mostly be formatter changes.

### CODE REVIEW

**Comment by chriselrod** [comment]
Documentation failure is because StrideArraysCore 0.4.7 wasn't available yet (package registry was slow to update).

**Comment by chriselrod** [comment]
Someone needs to review this before it can merge. I added a few people, so anyone who has time, please do.

**Comment by chriselrod** [comment]
This also includes a fix where in some cases it would use negative batch sizes, which would cause problems.

**Review by korsbo — APPROVED** [review]

File: src/loss.jl [hunk 1]
```diff
@@ -94,12 +90,12 @@ function chain_valgrad!(
     arg[i] = δ
     s += δ * δ
   end
-  return s, arg, pu
+  T(0.5) * s, arg, pu
```
Comment (andreuvall) [inline src/loss.jl:93]: This is intended, right?

Comment (chriselrod) [reply src/loss.jl:93]: Yes, it was weird that we were returning different values from valgrad and calling the NM directly with squared error.
The 0.5 means we don't need to multiply the gradient by 2.

**Pull request merged**